hoomd/operations.py

```diff
@@ -51,6 +116,17 @@ def _sys_init(self):
             return True
 
     def schedule(self):
+        """Prepares all operations for a `hoomd.Simulation.run` call.
+
+        This is provided in the public API to allow users to access quantities
```

Comment:
If it's possible to automatically call this before running, why isn't it automatically called when attempting to access a quantity that requires scheduling? Then this function wouldn't need to be in the public API at all, and it's entirely transparent to users. (There may be issues I'm not aware of with this suggestion, since I don't know what requires scheduling.)

Reply:
There are many quantities (i.e. the `kinetic_energy` property computed by `ThermoDynamicQuantities`) that are undefined when the operation is not scheduled (they return `None`). In the rare case that a user wants to compute these values before a call to `run`, they need to call `schedule` explicitly first. 

If these properties instead implicitly called `schedule`, one of the unintended side effects would be the resolution of all `TypeParameters` down C++. 1) This may lead to errors or warnings e.g. in pair.LJ parameters. 2) The types in the parameters are reduced to only those in the system. I think it would be awkward and confusing for users if a call to `thermo.kinetic_energy` resulted in errors from `pair.LJ`. For point 2, this potentially prevents use-cases like we employ in MoSDeF where we are using the `Logger` infrastructure to write out force field metadata, or write_debug_data. The output would depend on the order of items in the logger - e.g. if pair.LJ type parameters were written first they would have all types specified in the force field, but if the system potential energy is written first then the parameters will be collapsed.

Given the wide range of possible side effects, I think it best to make scheduling explicit for these cases and only implicit in the common case of `run`. Methods that require scheduling will be documented and raise an appropriate exception and we can clearly document all properties that return `None` before scheduling.

Reply:
Alternately, we could make schedule private and document these as being available after run. Users can call `sim.run(0)` to achieve this without advancing the system forward. `sim.run(0)` performs some extra steps like communication and integrator preparation that are not performed by `schedule` - see e.g. #726.

Reply:
@joaander I like your idea of `sim.run(0)`. I think that is conceptually easier to understand than the concept of "scheduling" (and explaining when it is or isn't necessary). I believe it is also closer to the current model of HOOMD 2?